### ISSUE TITLE
SslStream hangs on server cert with no private key

### ISSUE THREAD

**Issue by the-black-wolf** [dotnet/corefx#36438]
While testing out the core3 preview, I encountered a weird hang, I accidentally used loaded snakeoil pem key in X509Certificate (naturally without private key from /private but I realized that later), and tried to use it to authenticate as server. However, instead of throwing an exception to the fact that key is missing, or cert is invalid, the AuthenticateAsServer (also Async) hangs and both client and server are blocked. 

```
...
while (!tokenOfDeath.IsCancellationRequested)
    await server.AcceptTcpClientAsync()
        .ContinueWith(async tcpClient => await AcceptConnection(tcpClient.Result, tokenOfDeath));
...

private static async Task AcceptConnection(TcpClient tcpClient, CancellationToken token)
{
        var cert = X509Certificate.CreateFromCertFile("/etc/ssl/certs/ssl-cert-snakeoil.pem");
        var secure = new SslStream(tcpClient.GetStream(), false, 
                    (sender, certificate, chain, errors) => true, 
                    (sender, host, certificates, certificate, issuers) => cert);
        await secure.AuthenticateAsServerAsync(cert);
        // secure.AuthenticateAsServer(cert2); also hangs, not async issue
        Console.WriteLine("\x1b[34mServer::\x1b[0m Established secure connection...");
}
```
If I instantiate `X509Certificate2` from signed blob containing private key, it works like a charm. 

Expected result: exception indicating faulty certificate used.

**Comment by stephentoub** [comment]
Can you share a standalone repro?

**Comment by the-black-wolf** [comment]
Sure. Let me just recreate it as I moved on to other testing. I actually stumbled on another issue (also Linux related). When my client's protocol is set to Tls12 and server to Tls13 it also hangs in exactly the same manner (instead of throwing protocol mismatch)

**Comment by the-black-wolf** [comment]
Apologies for delay on this, I was busy last few days. I'll catch up weekend and provide a full example of both issues. I also noticed there is a newer preview of 3.0 so will also install the latest 3.0 preview to make sure it's still applicable.

**Comment by stephentoub** [comment]
Thanks.

**Comment by the-black-wolf** [comment]
I updated the question. The public/private key issue has disappeared, so either it was fixed in last preview or I did something wrong. However, the protocol mismatch hang on client is still present. I added code to issue.

**Comment by wfurt** [comment]
What version of Linux/OpenSSL do you use @the-black-wolf ? I can try to reproducer it and collect more info.

**Comment by the-black-wolf** [comment]
**mmix@BlackWolf:~$** lsb_release -a
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 18.10
Release:	18.10
Codename:	cosmic
**mmix@BlackWolf:~$** openssl version
OpenSSL 1.1.1  11 Sep 2018
**mmix@BlackWolf:~$** dotnet --version
3.0.100-preview3-010431
**mmix@BlackWolf:~$** dotnet --list-sdks
3.0.100-preview3-010431 [/opt/dotnet/sdk]

**Comment by wfurt** [comment]
I've seen some test failures with 1.1.1 and it may be related. There seems to be some behavior change we did not trace down yet. It may be interesting to try this on Ubuntu 18.04 if possible @the-black-wolf as that is LTS and has 1.1.0.

**Comment by the-black-wolf** [comment]
Nope, no joy. I installed Ubuntu Server LTS on vmware box and setup dotnet runtime and My store and it behaves the same (sorry for screenshot, I did not find a way to copy/paste text out of linux guest terminal). After the server dies on protocol missmatch, the client stays stuck. 

![image](https://user-images.githubusercontent.com/15175507/55405342-1dfc4380-555a-11e9-81a9-b6a8b0356326.png)

**Comment by the-black-wolf** [comment]
Hmmm, I just tried this on Windows, using the same preview, and it behaves exactly the same:

```
c:\projects\test>dotnet --list-sdks
2.1.502 [C:\Program Files\dotnet\sdk]
2.1.503 [C:\Program Files\dotnet\sdk]
2.1.504 [C:\Program Files\dotnet\sdk]
3.0.100-preview3-010431 [C:\Program Files\dotnet\sdk]

c:\projects\test>dotnet build
Microsoft (R) Build Engine version 16.0.443+g5775d0d6bb for .NET Core
Copyright (C) Microsoft Corporation. All rights reserved.

  Persisting no-op dg to c:\projects\test\obj\SslFail.csproj.nuget.dgspec.json
  Restore completed in 23.4 ms for c:\projects\test\SslFail.csproj.
C:\Program Files\dotnet\sdk\3.0.100-preview3-010431\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.RuntimeIdentifierInference.targets(151,5): message NETSDK1057: You are using a preview version of .NET Core. See: https://aka.ms/dotnet-core-preview [c:\projects\test\SslFail.csproj]
  SslFail -> c:\projects\test\bin\Debug\netcoreapp3.0\SslFail.dll

Build succeeded.
    0 Warning(s)
    0 Error(s)

Time Elapsed 00:00:01.04

c:\projects\test>dotnet run
Exception in Server: The client and server cannot communicate, because they do not possess a common algorithm.
```

So either this is broken across the board, or my sample is faulty in some way. Which is of course possible, but I am not seeing it. :(

**Comment by wfurt** [comment]
thanks for trying @the-black-wolf I was planning to give it try on Ubuntu 18.04 but I'm getting distracted with other issues.

**Comment by davidsh** [comment]
What happens if you reduce your repro to using .NET Core for the client (which you indicates hang on protocol mismatch error) and use another endpoint for the server?

// Requires TLS 1.2.
"https://www.ssllabs.com:10303/"

**Comment by the-black-wolf** [comment]
Ok, I have replaced `endPoint` with:
```csharp
var endPoint = new IPEndPoint(Dns.GetHostEntry("www.ssllabs.com").AddressList[0], 10303); 
```

and removed server and server `AutoResetEvent`. 

When I request Tls1.2 it negotiates ok (`All is well, no bug detected!`), when I demand Tls1.1 or 1.3, I get a protocol mismatch exception on client, as is supposed to 

`Exception in Client: Authentication failed, see inner exception. SSL Handshake failed with OpenSSL error - SSL_ERROR_SSL.`

So, apparently the .net core client only hangs when talking to .net core server.

**Comment by the-black-wolf** [comment]
I am not familiar with binary exchange technical details on Tls, but this all looks like AuthenticateAsServer fails to send an error condition back to client, who keeps on waiting for server to talk.

**Comment by davidsh** [comment]
>So, apparently the .net core client only hangs when talking to .net core server.

I think you might have introduced a deadlock in your code above when you use your own loopback server.  There is a lot of "Wait" type patterns in that code.  It's possible when the server is throwing an exception that it is doing it on the wrong thread (perhaps the main() thread) etc.

In fact lines like this in code:

```c#
await state.ServerStream.AuthenticateAsServerAsync(serverCert, false, SslProtocols.Tls13, false);
```

need to use .ConfigureAwait(false) to avoid these kinds of deadlocks.

**Comment by davidsh** [comment]
>but this all looks like AuthenticateAsServer fails to send an error condition back to client, who keeps on waiting for server to talk.

And if you think there is a bug in SslStream.AuthenticatAsServer, then it probably would be best to construct a repro using .NET Core for server-side but use .NET Framework for client-side.  Again, this is to isolate the repro better.  I still think there is a deadlock pattern in your code above.

**Comment by the-black-wolf** [comment]
The `.ConfigureAwait(false)` did nothing. But, I don't expect it though, I do catch the server side exception, so its not throwing outside and server does reach the end of its task. I don't think its a deadlock, but I'll give one quick test with old style threads and then try with framework client later.

**Comment by stephentoub** [comment]
In what situation would a server not close a connection after a failed SSL authentication? Isn't the connection basically unusable at that point?

**Comment by the-black-wolf** [comment]
not really, right? SSL/Tls is a tunneling protocol, you can still use high level connection if you like outside negotiated tunnel (STARTSSL works that way)

**Comment by davidsh** [comment]
>In what situation would a server not close a connection after a failed SSL authentication? Isn't the connection basically unusable at that point?

The server should close the socket after a failed TLS handshake.  Or at the very least, it is supposed to send back a TLS alert.

**Comment by the-black-wolf** [comment]
SslStream should never drop a connection by closing underlaying stream if `leaveInnerStreamOpen` is true. But again, even on false, it should notify client before dying (the www.ssllabs.com did). 
If the server has unilaterally stopped talking to client on mismatch then its a real bug.

**Comment by stephentoub** [comment]
> SslStream should never drop a connection by closing underlaying stream if leaveInnerStreamOpen is true

I'm not saying SslStream would itself close it, but generally a server (the code using SslStream) would.

> the www.ssllabs.com did

How? What did it send on the wire? How does that map or not map to the SSL spec.

**Comment by the-black-wolf** [comment]
> I'm not saying SslStream would itself close it, but generally a server (the code using SslStream) would.

That will happen eventually, but Tls has to complete its cycle so that both client and server exit the "secure" stage. Keeping one side in the dark is never a good practice nor is dropping connection without explanation. Thats  what douche firewalls do :)

> How? What did it send on the wire? How does that map or not map to the SSL spec?

No idea, as I said I am not familiar with the Tls protocol, so even WireSharking wouldnt mean much to me. Its just common logic. Server is obviously the one who decides that mismatch is occurring, client has no idea that things have gone bad on server side, for it to properly conclude (and return appropriate exception) it is obviously required that server sends something back before it dies.

**Comment by the-black-wolf** [comment]
OK, I completely removed async from code, I added both client and server as independent threads and the behavior is the same. I will try framework client tomorrow 

Source is here: https://pastebin.com/8BN23Auu

**Comment by stephentoub** [comment]
> The server should close the socket after a failed TLS handshake. Or at the very least, it is supposed to send back a TLS alert.

@davidsh, it would be SslStream that would send a TLS alert, right?  So does this mean we're missing that?

**Comment by davidsh** [comment]
>@davidsh, it would be SslStream that would send a TLS alert, right? So does this mean we're missing that?

Historically, SslStream has had problems in .NET Framework.  For example, we wouldn't send TLS CloseNotify alerts when closing the TLS handshake due to error.  And we had problems on the receiving side of SslStream is dealing with received TLS alerts.  Some of these problems were fixed in servicing updates of .NET Framework.

I don't remember where we ended up in .NET Core SslStream.  There might be some things we need to fix on the server-side and/or the client-side of SslStream regarding TLS alerts.

We need to tease apart the sending vs. receiving side of this repro and look at Wireshark traces to see what side might be at fault.

It is standard practice for most servers, for example, to close the Socket after a TLS handshake fails.  But there might be some server scenarios that choose to only send a TLS alert on a failed handshake.  And then resume normal TCP processing on that socket.  I don't remember what the TLS RFC says about that in practice.  I would need to investigate that.

**Comment by the-black-wolf** [comment]
That sounds like a probably cause of this. I'll try the framework client now, see what happens. 

As for debate, as I remember from the old days of implementing some RFCs, there was always a rule to not overdo things and be tolerant in implementation where it does not break it.  imho, unless Tls protocol by its nature leaves trash on sockets, there is no real need to force close connection on error, if C/S want to, they can do it themselves. One of the reasons I suspect this is the case is that whoever wrote SslStream on Framework introduced `leaveInnerStreamOpen` parameter on [SslStream Constructor](https://docs.microsoft.com/en-us/dotnet/api/system.net.security.sslstream.-ctor?view=netframework-4.7.2#System_Net_Security_SslStream__ctor_System_IO_Stream_System_Boolean_). Mandatory closure of connection would make this parameter faulty.

**Comment by wfurt** [comment]
You may try to get packet captures @the-black-wolf .
That would definitely show what side is doing what. 

I think if TLS handshake fails the method should throw so caller can deal with underlying stream. (and close as good  citizen would do)

**Comment by the-black-wolf** [comment]
Well, things are getting more interesting. While testing two-process connect on Windows (when testing core client <=> core server) requesting the same protocol (Tls1.3) the client (!) declared mismatch (?) and force dropped the connection (even though `leaveInnerStreamOpen` == true for both). When I switch them both to Tls 1.2, they negotiate properly, so its not my code. 
Do not know what to think of this, this sounds like a separate issue with Tls1.3 on Windows or SslStream issue with Tls1.3 on Windows. No idea why client and server would not connect on Tls1.3 on the same box, presumably the same cipher suits are available to both. When I let them auto-negotiate protocol they choose Tls1.2, so maybe Windows is not Tls1.3 ready?  (I am using Windows Server 2019)

On Linux both attempts with matching Tls negotiated properly. Auto-negotiation goes to Tls1.3.

Continuing testing.

**Comment by the-black-wolf** [comment]
I created a repo for my testing, you can for or use it for your own testing if you like. It makes it easier for me to cross-check windows and Linux on the same source. 

master branch contains Auto-negotiated server and client with embedded self-signed PFX (avoid any issues with certificates and stores). I'll create branches for various flavors of clients an servers and their protocols. 

https://github.com/the-black-wolf/SslFail

**Comment by the-black-wolf** [comment]
Ok, I switched client to Framework 4.7.2 and left server on core. It behaves exactly the same, client is stuck. Code is in https://github.com/the-black-wolf/SslFail/tree/frame-core-1.1-1.2 branch.

**Comment by the-black-wolf** [comment]
When I grab some time today or tomorrow, I'll try to wireshark it and see what comes out.

**Comment by davidsh** [comment]
>While testing two-process connect on Windows (when testing core client <=> core server) requesting the same protocol (Tls1.3) the client (!) declared mismatch (?). When I switch them both to Tls 1.2, they negotiate properly, so its not my code.

Windows doesn't fully support TLS 1.3 yet.  It has some support but only in the most recent Windows 10 OS and requires a registry key opt-in.  Plus it is not on the official RFC for TLS 1.3 yet. It is considered experimental right now.  It will be activated probably in the next release of Windows 10.

**Comment by the-black-wolf** [comment]
No problem, I suspected as much. Thats why I used 1.1 on 1.2 combo for framework client on core server. That should work regardless of Tls1.3 status.

**Comment by the-black-wolf** [comment]
OK, frame on frame also does not work, so this is an issue that persists from Framewok. 
https://github.com/the-black-wolf/SslFail/tree/frame-frame-1.1-1.2

I also created a core-eternal (ssllabs) branch, and that one properly throws on tls alert from server, reinfocing the idea that client is ok, but server fails to send tls alert. So I'll try to do comparative wire logs, and that's as much as I can help you with here. This is definitely something that should be fixed.

**Comment by the-black-wolf** [comment]
Ok, I did the wireshark, and its definitely TLS Alert issue. 

1. When calling ssllabs, it clearly sends TLS Alert message back before disconnecting (frame 7):
![image](https://user-images.githubusercontent.com/15175507/55617421-54bd9e00-5794-11e9-888f-7ac474a4aedb.png)
Trace file is [here](https://github.com/the-black-wolf/SslFail/blob/core-ssllabs-1.1-1.2/sslfailclient/log_ending_in_tlsalert.pcapng)

2. When doing core on core mismatch, the server does not send TLS Alert:
![image](https://user-images.githubusercontent.com/15175507/55618418-c1399c80-5796-11e9-94be-39228ffc8e21.png)
Trace file is [here](https://github.com/the-black-wolf/SslFail/blob/core-core-1.1-1.2/sslfailclient/log_without_tlsalert.pcapng)

so, obviously that is the difference which causes the client to hang waiting. Core server needs to send TlsAlert Fatal:70 on protocol mismatch.

**Comment by the-black-wolf** [comment]
Now that I knew what to look for I found a relevant entry in [RFC5246, section 7.2 ](https://tools.ietf.org/html/rfc5246#section-7.2) so you can find more info there. 

On an related not re forced TCP connection closure, the RFC states:
> No part of this standard should be taken to dictate the manner in which a usage profile for TLS manages its data transport, including when connections are opened or closed."

So you should be leaving the connection open if requested in `SslStream` constructor, AFAIS.

**Comment by davidsh** [comment]
@the-black-wolf 

>so, obviously that is the difference which causes the client to hang waiting. Core server needs to send TlsAlert Fatal:70 on protocol mismatch.

Thank you for the detailed analysis.  This is a bug in SslStream that appears to have existed even in .NET Framework. We will need to fix this on .NET Core.

**Comment by karelz** [comment]
@wfurt are you working on it? If not, please unassign yourself. Thanks!

**Comment by wfurt** [comment]
yes, I do @karelz. I have fix for OpenSSL but schanel gives me some grief.
